### ISSUE TITLE
Design suggestion: consider adding jax.random.Generator

### ISSUE THREAD

**Issue by NeilGirdhar** [google/jax#2294]
Edited in August 2020:

Jax's splittable RNGs are a really great idea!  However, it might be a good idea to expose the key as a class so that

- if Jax ever generalizes to other splittable RNGs, the change can be made transparently,
- the type annotation linters (like mypy) can verify key usage (currently DeviceArray is used for other things too),
- to make code a bit less unnecessarily verbose, and
- to be closer to the elegant numpy interface.

For example: [tjax.generator](https://github.com/NeilGirdhar/tjax/blob/master/tjax/generator.py).

**Comment by NeilGirdhar** [comment]
It might also help to have `jnp.random.default_rng(seed)`

**Comment by NeilGirdhar** [comment]
Looks like numpy already has splittable RNGs: https://numpy.org/devdocs/reference/random/parallel.html

Their design looks pretty outstanding, so it might be best to copy it.

**Comment by NeilGirdhar** [comment]
Also, rkern has a very insightful comment about the RNG used in Jax.  I don't understand all of it, but it seems fascinating.

https://github.com/numpy/numpy/issues/15656#issuecomment-592039584

**Comment by shoyer** [comment]
I agree that a class based interface for random number generation in JAX would be convenient. One reason why this could be a little tricky (vs the current function based approach) is that the class would need to be closed under JAX's transformation rules like batching.

I don't think we want stateful RNGs in JAX (these make functional transformations very hard), which rules out copying NumPy's API exactly. But maybe there's some aspect of the `SeedSequence` design or NumPy's new RNGs that could be helpful for us.

**Comment by NeilGirdhar** [comment]
@shoyer Thanks for the clear explanation.  That sounds really unfortunate—not just for random number generators, but for JAX objects in general.  It feels like we're going back 25 years to when C code would all be structures, and you'd have functions that accepted the structures as the first parameters.  Objects are superior to this pattern for many reason including inheritance and polymorphism.  It would be nice to keep those properties.

I'm not sure if it's possible, but I've opened an issue to discuss making that possible: https://github.com/google/jax/issues/2328.

**Comment by shoyer** [comment]
Classes work totally fine in JAX as long as you don't mutate state. Even your exact code example runs, e.g., see this notebook for a working example: 
https://gist.github.com/shoyer/0b3221ed0431befdfbfc9884e9353f8e

The reason why classes are awkward is that JAX's transforms like `vmap` don't necessarily work like you might expect if your functions return custom objects. For example, here's how you could auto-batch with JAX's functional interface:
```python
keys = jax.vmap(random.PRNGKey)(jax.numpy.arange(3))
samples = jax.vmap(random.uniform)(keys)
```
With a class, you need to write something like:
```
rngs = jax.vmap(rng)(jax.numpy.arange(3))
samples = jax.vmap(RNG.uniform)(rngs)
```
Which is fine, I guess, but not very convenient.

Intuitively, one might expect that batching the creation of a class would pre-batch all of its methods, but definitely doesn't work right now (`jax.vmap(rng)(jax.numpy.arange(3)).uniform()` results in an error) and I'm not even sure the behavior is entirely well defined.

**Comment by NeilGirdhar** [comment]
Just adding a comment in this thread, even though I gave a code example in the linked issue #2328.

I think a better approach would be to make the classes all optionally vectorized from the start, so the RNG class would accept an optional shape parameter too.  Then you don't need to vmap it, but rather do something like this:
```
rngs = RNG(jax.numpy.arange(3), jax_shape=(3,))
samples = rngs.uniform()
```
Would having this jax shape in all mutable objects allow us to do this?

Working with values and non-polymorphic functions is fairly annoying.

**Comment by mattjj** [comment]
Thanks for the ideas!

IIUC the main proposal is to have a distinguished PRNG key type, e.g. for mypy type checking. That sounds like a good idea! It wouldn't require any API additions or changes. In fact, in the [side-effects branch](https://github.com/google/jax/blob/side-effects/effects.py) we prototyped exactly that. We plan to merge something along those lines in the next couple months.

There are also ideas around API changes or internal refactorings, but I find those less compelling:
* We don't have any foreseeable-future plans to add other splittable PRNG hash functions yet, and IMO we shouldn't build anything motivated by that abstract goal until we know exactly the problem we're trying to solve.
* I don't think `rng, sub_rng = split(rng)` is more verbose than `rng, sub_rng = rng.split()`, and IMO by being more OOP-y the latter is less clear about PRNG key splitting having value semantics.
* Being closer to the NumPy PRNG interface is not a goal for JAX's PRNG, and NumPy's PRNG interface was not designed with the same considerations we have (especially around functional purity). Users who really want a NumPy interface might be able to build one on top.

Is my understanding of the main proposal (a distinguished PRNG key type, so keys aren't DeviceArrays but still work as they do now with JAX transformations) correct? If so, I propose we tweak the title of the issue, and leave it open until we land something along those lines.

WDYT?

**Comment by NeilGirdhar** [comment]
> IIUC the main proposal is to have a distinguished PRNG key type, e.g. for mypy type checking.

The full proposal is to have a class type for PRNGs.  You're right that one benefit is type-checking.  Another benefit is abstraction and the ability to call methods on objects, which enables polymorphism.  I think you may be underestimating he benefits of OO in writing clean code, but okay, let's cross that bridge when we get there.  I may soon have a more motivating example.

>  the side-effects branch we prototyped exactly that. We plan to merge something along those lines in the next couple months.

Awesome!  I don't totally understand it yet, but looking forward to it.

> IMO we shouldn't build anything motivated by that abstract goal until we know exactly the problem we're trying to solve.

Fair enough.

> I don't think rng, sub_rng = split(rng) is more verbose than rng, sub_rng = rng.split(), and IMO by being more OOP-y the latter is less clear about PRNG key splitting having value semantics.

Right, the only benefit is polymorphism.

**Comment by NeilGirdhar** [comment]
I've run into a problem with Jax using device arrays as the random number generator, which I've edited my issue with:

**An RNG passed to a jitted function will not induce recompilation of that jitted function**

Here's an example:
```
In [7]: @partial(jit, static_argnums=(0,))
   ...: def f(rng):
   ...:     print("JIT", rng)
   ...:     return 0.0

In [12]: a = PRNGKey(10)

In [13]: b = PRNGKey(10)

In [14]: f(a)
JIT [ 0 10]
Out[14]: 0.0

In [15]: f(b)
JIT [ 0 10]
Out[15]: 0.0
```

Because device arrays aren't hashable (rightfully so), then jit isn't able to find the static arguments in its cache (rightfully so).  What do you think of this generator type, which I've been using in my own code?
```
from __future__ import annotations

from typing import Any, List, Optional, Tuple

import jax.random
from jax import numpy as jnp

from .annotations import Tensor
from .tensor_like import TensorLike

__all__ = ['Generator']


class Generator(TensorLike):

    def __init__(self,
                 *,
                 seed: Optional[int] = None,
                 key: Optional[Any] = None,
                 **kwargs):
        super().__init__(**kwargs)
        self.key = (key
                    if key is not None
                    else jax.random.PRNGKey(seed))

    # New methods -------------------------------------------------------------
    def split(self, n=2) -> List[Generator]:
        keys = jax.random.split(self.key, n)
        return [Generator(key=key) for key in keys]

    def __repr__(self):
        return f"{type(self).__name__}({self.key})"

    # Implemented methods -----------------------------------------------------
    @classmethod
    def from_tensor(cls, values, aux_data):
        return cls(key=values)

    def to_tensor(self):
        return self.key, None

    # Magic methods -----------------------------------------------------------
    def __eq__(self, other):
        if not isinstance(other, Generator):
            return NotImplemented
        return jnp.all(self.key == other.key)

    def __ne__(self, other):
        if not isinstance(other, Generator):
            return NotImplemented
        return not self.__eq__(other)

    def __hash__(self):
        return hash((int(self.key[0]), int(self.key[1])))
```

**Comment by shoyer** [comment]
@NeilGirdhar the value returned by `jax.random.PRNGKey` is just an JAX array (though the exact form is an implementation detail), so the simple way to avoid recompilation is to avoid using `static_argnums`.

**Comment by NeilGirdhar** [comment]
@shoyer That's true, but unfortunately in my case the RNG is part of a larger object that needs to be passed in as a static argument.

**Comment by NeilGirdhar** [comment]
@shoyer After some more thought, you're right.  I should avoid using static arguments.

**Comment by NeilGirdhar** [comment]
(Closing because Haiku appears to have a more elegant solution.)
